pkg/controller/user-cluster-controller-manager/constraint-syncer/controller.go

```diff
@@ -145,7 +112,26 @@ func (r *reconciler) reconcile(ctx context.Context, constraint *kubermaticv1.Con
 	if err := reconciling.ReconcileUnstructureds(ctx, constraintCreatorGetters, "", r.userClient); err != nil {
 		return fmt.Errorf("failed to reconcile constraint: %v", err)
 	}
+	log.Debugw("constraint created")
+	return nil
+}
 
+func (r *reconciler) cleanupConstraint(ctx context.Context, constraint *kubermaticv1.Constraint, log *zap.SugaredLogger) error {
+	log = log.With("constraint", constraint)
+	log.Debugw("cleanup processing:")
+
+	toDelete := &unstructured.Unstructured{}
+	toDelete.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   constrainthandler.ConstraintsGroup,
+		Version: constrainthandler.ConstraintsVersion,
+		Kind:    constraint.Spec.ConstraintType,
+	})
+	toDelete.SetName(constraint.Name)
+
+	if err := r.userClient.Delete(ctx, toDelete); err != nil && !kerrors.IsNotFound(err) {
+		return fmt.Errorf("failed to delete constraint: %v", err)
+	}
```

Comment:
Why aren't you using `constraint` directly?

Reply:
ok, I think I got it. The resource you are deleting is not the kubermatic `Constraint`.

Reply:
coz we are creating/deleting unstructured gatekeeper Constraint object in user-cluster with `constraints.gatekeeper.sh/v1beta1`

Reply:
@lsviben yeah, but is there any reason why we are not using the structured version? Is it to avoid an import?

Reply:
The Constraint CRDs for Gatekeeper are dynamic so we need to use unstructured because we never know what we get. 

Gatekeeper-controller creates the CRD for Constraints depending on the Constraint Template the user creates: https://open-policy-agent.github.io/gatekeeper/website/docs/howto/#constraint-templates

Reply:
ok, interesting. Thanks for clarifying.